### PR TITLE
mlflow logging integration with yolox  training

### PR DESCRIPTION
**Pull request by Im-Himanshu**
Needed integration of ``yolox`` to log experiments with mlflow, the pull request provide additional option in -l --logger argument to log output in "mlflow".

Requires an environment file (``.env``) in the root folder of the projects.

required additional dependency of ``mlflow`` and python-dotenv failing which error is raised if logger is set to mlflow.

### CODE REVIEW

**Comment by Im-Himanshu** [comment]
Tested the logging on to data bricks, following logs are available for all the runs.
image

**Comment by Im-Himanshu** [comment]
@FateScript Requesting you to please review the pull request.

File: yolox/core/trainer.py [hunk 1]
```diff
@@ -198,6 +202,8 @@ def after_train(self):
         if self.rank == 0:
             if self.args.logger == "wandb":
                 self.wandb_logger.finish()
+            if self.args.logger == "mlflow":
```
Comment (FateScript) [inline yolox/core/trainer.py:205]: `elif` instead of `if`

File: yolox/utils/logger.py [hunk 2]
```diff
@@ -438,3 +438,329 @@ def initialize_wandb_logger(cls, args, exp, val_dataset):
...
+import importlib.util
+from collections.abc import MutableMapping
+import packaging.version
+import json, platform, time
+
+ENV_VARS_TRUE_VALUES = {"1", "ON", "YES", "TRUE"}
+
+
+
+class MlflowLogger(object):
```
Comment (FateScript) [inline yolox/utils/logger.py:452]: please rm object here, it's useless.

File: yolox/utils/logger.py [hunk 3]
```diff
@@ -438,3 +438,329 @@ def initialize_wandb_logger(cls, args, exp, val_dataset):
                     wandb_params.update({k[len(prefix):]: v})
 
         return cls(config=vars(exp), val_dataset=val_dataset, **wandb_params)
+
+import importlib.metadata
```
Comment (FateScript) [inline yolox/utils/logger.py:442]: The logger file might be too large, I highly recommend to use a new file named `mlflow_logger.py`.

File: yolox/core/trainer.py [hunk 4]
```diff
@@ -276,6 +282,12 @@ def after_iter(self):
                         "train/lr": self.meter["lr"].latest
                     })
                     self.wandb_logger.log_metrics(metrics, step=self.progress_in_iter)
+                if self.args.logger=='mlflow':
+                    logs = {"train/" + k: v.latest for k, v in loss_meter.items()}
+                    logs.update({
```
Comment (FateScript) [inline yolox/core/trainer.py:287]: Is there any need to break the line here?

File: yolox/utils/logger.py [hunk 5]
```diff
@@ -438,3 +438,329 @@ def initialize_wandb_logger(cls, args, exp, val_dataset):
...
+        self._mlflow_log_artifacts = os.getenv("YOLOX_MLFLOW_LOG_MODEL_ARTIFACTS", "False").upper() in ENV_VARS_TRUE_VALUES
+        self._mlflow_log_model_per_n_epochs = int(os.getenv("YOLOX_MLFLOW_LOG_MODEL_PER_n_EPOCHS", 30))
+        self._mlflow_log_nth_epoch_models = os.getenv("YOLOX_MLFLOW_LOG_Nth_EPOCH_MODELS", "False").upper() in ENV_VARS_TRUE_VALUES
+        self.run_name = os.getenv("YOLOX_MLFLOW_RUN_NAME", None)
+        self.run_name = None if len(self.run_name.strip())==0 else self.run_name
+        self._flatten_params = os.getenv("YOLOX_MLFLOW_FLATTEN_PARAMS", "FALSE").upper() in ENV_VARS_TRUE_VALUES
+
+        self._nested_run = os.getenv("MLFLOW_NESTED_RUN", "FALSE").upper() in ENV_VARS_TRUE_VALUES
+        self._run_id = os.getenv("MLFLOW_RUN_ID", None)
+
```
Comment (FateScript) [inline yolox/utils/logger.py:556]: plz delete this empty line here. Two lines of empty in a function is not a good coding style.

File: yolox/utils/logger.py [hunk 6]
```diff
@@ -438,3 +438,329 @@ def initialize_wandb_logger(cls, args, exp, val_dataset):
...
+        # https://github.com/mlflow/mlflow/pull/9705
+        # https://github.com/mlflow/mlflow/releases/tag/v2.8.0
+        self._async_log = packaging.version.parse(self._ml_flow.__version__) >= packaging.version.parse("2.8.0")
+
+        logger.debug(
+            f"MLflow experiment_name={self._experiment_name}, run_name={self.run_name}, nested={self._nested_run},"
+            f" tags={self._nested_run}, tracking_uri={self._tracking_uri}"
+        )
+        rank = trainer.rank
+        if rank==0:
```
Comment (FateScript) [inline yolox/utils/logger.py:568]: Please use `is_main_process` in dist.py.

UPDATE:  so many `rank == 0`  below, please fix them all.

File: yolox/core/trainer.py [hunk 7]
```diff
@@ -387,4 +410,4 @@ def save_ckpt(self, ckpt_name, update_best_ckpt=False, ap=None):
                         "best_ap": self.best_ap,
                         "curr_ap": ap
                     }
-                )
+                )
```
Comment (FateScript) [inline yolox/core/trainer.py:413]: missing new line here.

File: yolox/utils/logger.py [hunk 8]
```diff
@@ -438,3 +438,329 @@ def initialize_wandb_logger(cls, args, exp, val_dataset):
...
+            for k, v in d.items():
+                key = str(parent_key) + delimiter + str(k) if parent_key else k
+                if v and isinstance(v, MutableMapping):
+                    yield from self.flatten_dict(v, key, delimiter=delimiter).items()
+                else:
+                    yield key, v
+
+        return dict(_flatten_dict(d, parent_key, delimiter))
+
+    def setup(self, trainer):
```
Comment (FateScript) [inline yolox/utils/logger.py:497]: I notice that only trainer.exp and trainer.args is needed here. Should we use `exp, args` instead of `trainer`?

UPDATE: If you make sure that you might use some attributes in trainer in the future, please use `trainer` as args. However, if you make sure that you will not use more attrs in the future, some specific arguments is better.
This princple holds for all the following methods.

**Review by FateScript — CHANGES_REQUESTED** [review]
@Im-Himanshu Thanks for your contribution : )

Please check my review suggestion and lint your code.

File: yolox/core/trainer.py (hunk 4 above)
Comment (Im-Himanshu) [reply yolox/core/trainer.py:287]: I copied these two line from wandb logger, I think it is output of code beautifier, any removed the extra line.

File: yolox/utils/logger.py (hunk 6 above)
Comment (Im-Himanshu) [reply yolox/utils/logger.py:568]: replaced them all with is_main_process()

File: yolox/utils/logger.py (hunk 3 above)
Comment (Im-Himanshu) [reply yolox/utils/logger.py:442]: created a new file and shifted the code there

**Comment by FateScript** [comment]
Any update? @Im-Himanshu

File: yolox/utils/logger.py (hunk 8 above)
Comment (Im-Himanshu) [reply yolox/utils/logger.py:497]: removed trainer as input from all function and further chunked it into exact dependency.

**Comment by Im-Himanshu** [comment]
> Any update? @Im-Himanshu

@FateScript Pardon me for the delayed response, I have pushed new commits to implement all the suggestions.
Kindly review.

**Comment by Im-Himanshu** [comment]
> > Any update? @Im-Himanshu
> 
> @FateScript Excuse me for the delayed response, I have pushed new commits to implement all the suggestions. Kindly review.

@FateScript Gentle Reminder to Please check and merge the request.

File: docs/mlflow_integration.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+### Step-3: Set up MLFlow Environment Variables
+Set the following environment variables in your `.env` file:
+```bash
+MLFLOW_TRACKING_URI="127.0.0.1:5000"  # set to your mlflow server URI
+MLFLOW_EXPERIMENT_NAME="/path/to/experiment"  # set to your experiment name
+MLFLOW_TAGS={"release.candidate": "DEV1", "release.version": "0.0.0"}
+# config related to logging model to mlflow as pyfunc
+YOLOX_MLFLOW_LOG_MODEL_ARTIFACTS="True" # whether to log model or not 
+YOLOX_MLFLOW_LOG_HISTORICAL_MODEL_PER_n_EPOCHS=30 # log model after every n epochs
```
Comment (Cloudhax23) [inline docs/mlflow_integration.md:26]: Should be YOLOX_MLFLOW_LOG_MODEL_PER_n_EPOCHS

File: docs/mlflow_integration.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+YOLOX_MLFLOW_LOG_HISTORICAL_EPOCH_MODELS="False" # whether to log step model along with best_model or not
```
Comment (Cloudhax23) [inline docs/mlflow_integration.md:27]: Should be YOLOX_MLFLOW_LOG_Nth_EPOCH_MODELS

**Review by Cloudhax23 — CHANGES_REQUESTED** [review]

File: yolox/utils/logger.py [hunk 11]
```diff
@@ -98,8 +98,11 @@ def setup_logger(save_dir, distributed_rank=0, filename="log.txt", mode="a"):
 
     logger.remove()
     save_file = os.path.join(save_dir, filename)
+    crnt_log_save_file = os.path.join(save_dir, 'train_log_crnt.txt')
```
Comment (FateScript) [inline yolox/utils/logger.py:101]: Why this `train_log_crnt.txt` is needed?  Seems that your code redirect io to it and remove this file if it exists.

File: docs/mlflow_integration.md (hunk 9 above)
Comment (FateScript) [reply docs/mlflow_integration.md:26]: @Im-Himanshu  Please update your doc

File: docs/mlflow_integration.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+# one working example is this
+python tools/train.py -l mlflow -f exps/example/custom/yolox_s.py -d 1 -b 8 --fp16 -o -c pre_trained_model/yolox_s.pth
+```
+### Step-4: optional; start the mlflow ui and track your experiments
+If you log runs to a local mlruns directory, run the following command in the directory above it, then access http://127.0.0.1:5000 in your browser.
+
+```bash
+mlflow ui --port 5000
+```
+
```
Comment (FateScript) [inline docs/mlflow_integration.md:50]: Two many blank line here.

File: yolox/utils/mlflow_logger.py [hunk 13]
```diff
@@ -0,0 +1,373 @@
+#!/usr/bin/env python3
+# Copyright (c) Megvii Inc. All rights reserved.
+# Please read docs/mlflow_integration.md for more details.
+
+import json
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:5]: `isort`/ `ruff check --fix` this file(and your update) please

File: yolox/utils/mlflow_logger.py [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+                        dict(combined_dict_items[i: i + self._MAX_PARAMS_TAGS_PER_BATCH]),
+                        synchronous=False
+                    )
+                else:
+                    self._ml_flow.log_params(
+                        dict(combined_dict_items[i: i + self._MAX_PARAMS_TAGS_PER_BATCH])
+                    )
+
+    def convert_exp_todict(self, exp):
+        """
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:232]: This project use google style docstring.
For example
```
decription of function/method

Args:
     arg (type): description

Returns:
```
Please update your docstring in a unified format. Thanks : )

File: yolox/utils/mlflow_logger.py [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+                    )
+
+            if self._async_log:
+                self._ml_flow.log_metrics(metrics=metrics, step=step, synchronous=False)
+            else:
+                self._ml_flow.log_metrics(metrics=metrics, step=step)
+
+    def on_train_end(self, args, file_name, metadata):
+        """
+        Input:
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:283]: Google style, please

File: yolox/utils/mlflow_logger.py [hunk 16]
```diff
@@ -0,0 +1,373 @@
...
+                       'test_conf', 'nmsthre']
+        exp_dict = {k: v for k, v in exp.__dict__.items()
+                    if not k.startswith("__") and k in filter_keys}
+        return exp_dict
+
+    def on_log(self, args, exp, step, logs):
+        """
+        Log metrics to MLflow.
+
+        Input:
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:252]: Google style, ditto.

File: yolox/utils/mlflow_logger.py [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+            args: training args dictionary
+            file_name: output directory
+            metadata: model related metadata
+        mlflow logging action to take when training ends:
+
+        1. log the training log file
+        2. publish the latest best model to model_registry if it is allowed in config file
+        3. close the mlfow run
+
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:292]: Two many blank lines here.

**Review by FateScript — CHANGES_REQUESTED** [review]
@Im-Himanshu  Please fix it.

**Comment by FateScript** [comment]
@Im-Himanshu  Also don't forget to lint your code.

File: yolox/utils/logger.py (hunk 11 above)
Comment (Im-Himanshu) [reply yolox/utils/logger.py:101]: @FateScript As you have suggested in your earlier review that you recommend creating a new log file to be uploaded. it is deleted at that start to keep only current run logs in this.
image

File: yolox/utils/mlflow_logger.py (hunk 13 above)
Comment (Im-Himanshu) [reply yolox/utils/mlflow_logger.py:5]: image
 Done

**Comment by Im-Himanshu** [comment]
> @Im-Himanshu Also please don't forget to lint your code.

Linted the code, the only major issue in lint, is the import statement which has to be done inside class (same as being done in wandb logger) because this is an optional feature.
image

File: docs/mlflow_integration.md (hunk 9 above)
Comment (Im-Himanshu) [reply docs/mlflow_integration.md]: Done

File: yolox/utils/mlflow_logger.py [hunk 18]
```diff
@@ -0,0 +1,419 @@
...
+import importlib.util
+import json
+import os
+from collections.abc import MutableMapping
+import packaging.version
+
+from loguru import logger
+import torch
+from yolox.utils import is_main_process
+class MlflowLogger:
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:26]: Did you lint with this repo's setting correctly?  It should be two blank line before the class  definition.

You might check your lint result here: https://github.com/Megvii-BaseDetection/YOLOX/actions/runs/9796205233/job/27068037255?pr=1773

**Review by FateScript — COMMENTED** [review]
PTAL @Im-Himanshu

File: yolox/utils/logger.py (hunk 11 above)
Comment (Im-Himanshu) [reply yolox/utils/logger.py]: Removed it and reverted to old code.

though I think it would have been better to log only the current experiment logs to mlflow.

File: yolox/utils/mlflow_logger.py (hunk 18 above)
Comment (Im-Himanshu) [reply yolox/utils/mlflow_logger.py:28]: Yes, linted again to remove these lint issue.

**Comment by Im-Himanshu** [comment]
@FateScript Removed the additional logger and linted the code again to remove build process error.

**Comment by FateScript** [comment]
@Im-Himanshu  please isort your code(`isort -rc`), see [here](https://github.com/Megvii-BaseDetection/YOLOX/actions/runs/9806832237/job/27120887168?pr=1773) for more details.

**Comment by Im-Himanshu** [comment]
> @Im-Himanshu please isort your code(`isort -rc`), see [here](https://github.com/Megvii-BaseDetection/YOLOX/actions/runs/9806832237/job/27120887168?pr=1773) for more details.

in trainer.py and mlflow_logger.py, isort has transferred the ```import datetime``` at the end but it was not the case in original code, as of now I have manually shifted them to the top, please see if it is correct or not.

or if  there are any settings of isort that I can follow for this project, I can rerun the isort.

File: yolox/utils/mlflow_logger.py [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+import packaging.version
+from collections.abc import MutableMapping
+from loguru import logger
+
+import torch
+
+from yolox.utils import is_main_process
+
+
+
```
Comment (FateScript) [inline yolox/utils/mlflow_logger.py:29]: Two many blank lines here.

**Review by FateScript — CHANGES_REQUESTED** [review]

**Comment by FateScript** [comment]
> > @Im-Himanshu please isort your code(`isort -rc`), see [here](https://github.com/Megvii-BaseDetection/YOLOX/actions/runs/9806832237/job/27120887168?pr=1773) for more details.
> 
> @FateScript in `trainer.py` and `mlflow_logger.py`, `isort` has transferred the `import datetime` and other default python package import at the end but it was not the case in original code so as of now I have manually shifted them to the top, please see if it is correct or not.
> 
> I think my isort setting are not matched to the project setting, if there are any settings of isort that I can follow for this project, I can rerun the isort or otherwise the sort that I have done now should be correct based on my intuition.

If you instal the correct isort version(4.3.21 here), everything will be ok.
BTW, be sure of the format check [here](https://github.com/Megvii-BaseDetection/YOLOX/blob/main/.github/workflows/format_check.sh) passed before you commit.

**Comment by Im-Himanshu** [comment]
> > > @Im-Himanshu please isort your code(`isort -rc`), see [here](https://github.com/Megvii-BaseDetection/YOLOX/actions/runs/9806832237/job/27120887168?pr=1773) for more details.
> > 
> > 
> > @FateScript in `trainer.py` and `mlflow_logger.py`, `isort` has transferred the `import datetime` and other default python package import at the end but it was not the case in original code so as of now I have manually shifted them to the top, please see if it is correct or not.
> > I think my isort setting are not matched to the project setting, if there are any settings of isort that I can follow for this project, I can rerun the isort or otherwise the sort that I have done now should be correct based on my intuition.
> 
> If you instal the correct isort version(4.3.21 here), everything will be ok. BTW, be sure of the format check [here](https://github.com/Megvii-BaseDetection/YOLOX/blob/main/.github/workflows/format_check.sh) passed before you commit.

@FateScript My isort version was not as required by the library, hence the previous errors. 
Anyway I have removed the given pylint and isort errors. It should pass all the checks this time.

File: yolox/utils/__init__.py [hunk 20]
```diff
@@ -10,6 +10,7 @@
 from .ema import *
 from .logger import WandbLogger, setup_logger
 from .lr_scheduler import LRScheduler
+from .mlflow_logger import MlflowLogger
```
Comment (FateScript) [inline yolox/utils/__init__.py:13]: The checker got the following error: `yolox/utils/__init__.py Imports are incorrectly sorted.`
since `ml` is greater than `me` and smaller than `mo`, this line should be put after `.metric`.
make sure `isort -rc yolox` works.

Comment (Im-Himanshu) [reply yolox/utils/__init__.py]: updated, checked isort on whole package this time.

**Comment by FateScript** [comment]
LGTM.
